File: ruby/cut.rb
# intersection of two triangles

require "intersection"

class Cut

 attr_reader :intersection0, :intersection1

 def Cut.between(triangle0, triangle1)
  return nil if exclude?(triangle0, triangle1)
  intersections = Array.new
  triangle0.segments.each do |segment|
   intersection = Intersection.of(triangle1, segment)
   intersections << intersection if intersection
  end
  triangle1.segments.each do |segment|
   intersection = Intersection.of(triangle0, segment)
   intersections << intersection if intersection
  end
  return nil if 0 == intersections.size
  raise "improper cut #{intersections.size}" unless 2 == intersections.size
  Cut.new(intersections[0],intersections[1], triangle0, triangle1)
 end
 
 def Cut.exclude?(triangle0, triangle1)
  t0 = triangle0.node0
  t1 = triangle0.node1
  t2 = triangle0.node2

  s0 = triangle1.node0
  s1 = triangle1.node1
  s2 = triangle1.node2

  v0 = Intersection.volume6(t0,t1,t2,s0)
  v1 = Intersection.volume6(t0,t1,t2,s1)
  v2 = Intersection.volume6(t0,t1,t2,s2)
  return true if ( v0 > 0.0 && v1 > 0.0 && v2 > 0.0)
  return true if ( v0 < 0.0 && v1 < 0.0 && v2 < 0.0)

  v0 = Intersection.volume6(s0,s1,s2,t0)
  v1 = Intersection.volume6(s0,s1,s2,t1)
  v2 = Intersection.volume6(s0,s1,s2,t2)
  return true if ( v0 > 0.0 && v1 > 0.0 && v2 > 0.0)
  return true if ( v0 < 0.0 && v1 < 0.0 && v2 < 0.0)

  false
 end

 def initialize(intersection0, intersection1, triangle0, triangle1)
  @intersection0 = intersection0
  @intersection1 = intersection1
  @triangles = [triangle0, triangle1]
  triangle0.cuts << self
  triangle1.cuts << self
 end

end

File: src/cut.c

/* cut between two Triangles of two Intersections */

/* Copyright 2007 United States Government as represented by the
 * Administrator of the National Aeronautics and Space
 * Administration. No copyright is claimed in the United States under
 * Title 17, U.S. Code.  All Other Rights Reserved.
 *
 * The knife platform is licensed under the Apache License, Version
 * 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include <stdlib.h>
#include <stdio.h>
#include "cut.h"

#define TRY(fcn,msg)					      \
  {							      \
    int code;						      \
    code = (fcn);					      \
    if (KNIFE_SUCCESS != code){				      \
      printf("%s: %d: %d %s\n",__FILE__,__LINE__,code,(msg)); \
      return code;					      \
    }							      \
  }

#define cut_gather_intersection						\
  if ( NULL != intersection ) {						\
    if ( NULL == intersection0 ) { intersection0 = intersection; }	\
    else { if ( NULL == intersection1 ) { intersection1 = intersection; } \
      else { printf("%s: %d: cut_between improper intersection >2\n",	\
		    __FILE__,__LINE__); return KNIFE_IMPROPER; } } }

KNIFE_STATUS cut_establish_between( Triangle triangle0, Triangle triangle1 )
{
  Cut cut;
  Intersection intersection;
  Intersection intersection0, intersection1;
  int segment_index;

  if ( NULL == triangle0 || NULL == triangle1 ) return KNIFE_NULL;

  intersection  = NULL;
  intersection0 = NULL;
  intersection1 = NULL;

  for (segment_index = 0 ; segment_index < 3; segment_index++ )
    {
      TRY( intersection_of( triangle1, 
			    triangle_segment( triangle0, segment_index ),
			    &intersection ),
	   "triangle1 segment intersection" );
      cut_gather_intersection;
    }
  
  for (segment_index = 0 ; segment_index < 3; segment_index++ )
    {
      TRY( intersection_of( triangle0, 
			    triangle_segment( triangle1, segment_index ),
			    &intersection ),
	   "triangle0 segment intersection" );
      cut_gather_intersection;
    }

  if ( NULL != intersection0 && NULL == intersection1 )
    { 
      printf("%s: %d: cut_between improper intersection = 1\n",
	     __FILE__,__LINE__); 
      triangle_tecplot(triangle0);
      triangle_tecplot(triangle1);
      return KNIFE_IMPROPER; 
    }

  if ( NULL != intersection0 && NULL != intersection0 )
    {
      cut = (Cut)malloc( sizeof(CutStruct) );
      if (NULL == cut) 
	{
	  printf("%s: %d: malloc failed in cut_between\n",
		 __FILE__,__LINE__);
	  return KNIFE_MEMORY; 
	}
      
      cut->triangle0 = triangle0;
      cut->triangle1 = triangle1;

      TRY( triangle_add_cut( triangle0, cut ), "add cut to tri0");
      TRY( triangle_add_cut( triangle1, cut ), "add cut to tri1");

      cut->intersection0 = intersection0;
      cut->intersection1 = intersection1;
    }

  return KNIFE_SUCCESS; 
}

void cut_free( Cut cut )
{
  if ( NULL == cut ) return;
  free( cut );
}

